File: mpp/Lab9/javaPb9/persistence/src/main/java/ro/mpp2024/repository/RezultatRepository.java
package ro.mpp2024.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.model.Categorie;
import ro.mpp2024.model.Participant;
import ro.mpp2024.model.Proba;
import ro.mpp2024.model.Rezultat;
import ro.mpp2024.repository.interfaces.IRezultatRepository;
import ro.mpp2024.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class RezultatRepository implements IRezultatRepository {

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public RezultatRepository(JdbcUtils dbUtils) {
        logger.info("Initializing RezultatRepository with properties: {} ");
        this.dbUtils = dbUtils;
    }



    @Override
    public Rezultat createEntityFromResultSet(ResultSet resultSet) throws SQLException {
        logger.traceEntry("createEntityFromResultSet");
        Long id = resultSet.getLong("id");
        Long id_participant = resultSet.getLong("id_participant");
        String id_proba = resultSet.getString("id_proba");
        Long numarPuncte = resultSet.getLong("numar_puncte");

        Proba proba = null;
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Proba where id=?")){
            preStmt.setString(1,id_proba);
            try(ResultSet result=preStmt.executeQuery()){
                System.out.println(result);
                if(result.next()){
                    proba = new Proba(result.getString("id"),result.getString("nume")
                            , Categorie.valueOf(result.getString("categorie")));
                }
            }
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        Participant participant = null;
        try(PreparedStatement preStmt=con.prepareStatement("select * from Participant where id=?")){
            preStmt.setLong(1,id_participant);
            try(ResultSet result=preStmt.executeQuery()) {
                if (result.next()) {
                    participant = new Participant(result.getLong("id"), result.getString("nume")
                            , result.getString("prenume"), result.getInt("varsta"));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }

        Rezultat rezultat = new Rezultat(id,participant,proba, numarPuncte);
        logger.traceExit(rezultat);
        return rezultat;
    }


    @Override
    public Iterable<Rezultat> findAll() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Rezultat> rezultate = new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Rezultat")){
            try(ResultSet result=preStmt.executeQuery()){
                while(result.next()){
                    Rezultat rezultat = createEntityFromResultSet(result);
                    rezultate.add(rezultat);
                }
            }
        }catch (Exception ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit(rezultate);
        return rezultate;
    }

    @Override
    public Rezultat findOne(Long aLong) {
        logger.traceEntry("findOne");
        Connection con = dbUtils.getConnection();
        Rezultat rezultat = null;
        try(PreparedStatement preStmt=con.prepareStatement("select * from Rezultat where id=?")){
            preStmt.setLong(1,aLong);
            try(ResultSet result=preStmt.executeQuery()){
                if(result.next()){
                    rezultat = createEntityFromResultSet(result);
                }
            }
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(rezultat);
        return rezultat;
    }

    @Override
    public void save(Rezultat entity) {
        logger.traceEntry("save");
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Rezultat(id_proba,id_participant,numar_puncte) values (?,?,?)")){
            preStmt.setLong(2,entity.getParticipant().getId());
            preStmt.setString(1,entity.getProba().getId());
            preStmt.setLong(3,entity.getNumar_puncte());
            int result = preStmt.executeUpdate();
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public Rezultat delete(Long aLong) {
        // TODO: implement method when neeeded
        return null;
    }

    @Override
    public Rezultat update(Long aLong, Rezultat new_entity) {
        // TODO: implement method when neeeded
        return null;
    }

    @Override
    public Map<Participant, Long> listaParticipantiPuncteAlfabetic() {
        logger.traceEntry("listaParticipantiPuncteAlfabetic");
        Map<Participant, Long> map = new LinkedHashMap<>();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement presmt = con.prepareStatement("SELECT sum(r.numar_puncte) as total_punct, p.id, p.nume,p.prenume, p.varsta " +
                "FROM Rezultat as r Inner join Participant as p " +
                "On r.id_participant = p.id group by r.id_participant " +
                "order by lower(p.nume),lower(p.prenume)")){
            try(ResultSet result = presmt.executeQuery()){
                while(result.next()){
                    Participant participant = new Participant(result.getLong("id"),result.getString("nume"),result.getString("prenume"),result.getInt("varsta"));
                    map.put(participant,result.getLong("total_punct"));
                }
            }
        }
        catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceEntry("Terminat dictionar");
        return map;
    }

    @Override
    public Map<Participant, Long> listaParticiapntPuncteProbaDescrescator(String id_proba) {
        logger.traceEntry("listaParticiapntPuncteProbaDescrescator");
        Map<Participant, Long> map = new LinkedHashMap<>();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement presmt = con.prepareStatement("SELECT SUM(r.numar_puncte) AS total_puncte, p.id, p.nume, p.prenume, p.varsta " +
                "FROM Rezultat AS r INNER JOIN Participant AS p ON r.id_participant = p.id " +
                "WHERE r.id_proba = ? GROUP BY r.id_participant ORDER BY total_puncte DESC;")){
            presmt.setString(1,id_proba);
            try(ResultSet result = presmt.executeQuery()){
                while(result.next()){
                    Participant participant = new Participant(result.getLong("id"),result.getString("nume"),result.getString("prenume"),result.getInt("varsta"));
                    map.put(participant,result.getLong("total_puncte"));
                }
            }
        }
        catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceEntry("Terminat dictionar");
        return map;
    }

}

File: mpp/Lab12/csharpPb9/persistene/repository/RezultatRepository.cs
using csharpPb9.utils;
using log4net;
using System.Data;

public class RezultatRepository : IRezultatRepository
{
    private static readonly ILog log = LogManager.GetLogger("");

    IDictionary<string, string> props;

    public RezultatRepository(IDictionary<string, string> props)
    {
        log.Info("Creating ProbaRepository with properties: " + props);
        this.props = props;
    }

    public Rezultat fromSettoEntity(IDataReader read)
    {
        long id = read.GetInt64(0);
        long id_participant = read.GetInt64(2);
        string id_proba = read.GetString(1);
        int scor = read.GetInt32(3);

        #region Find participant and proba
        var con = DBUtils.getConnection(props);
        Participant p = null;
        try
        {
            log.Info("Finding participant");
;
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT * FROM Participant WHERE id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id_participant;
                comm.Parameters.Add(paramId);
                using (IDataReader dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        p = new Participant(dataR.GetInt64(0), dataR.GetString(1), dataR.GetString(2), dataR.GetInt32(3));
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }

        con = DBUtils.getConnection(props);
        Proba proba = null;
        try
        {
            log.Info("Finding proba");
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT * FROM Proba WHERE id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id_proba;
                comm.Parameters.Add(paramId);
                using (IDataReader dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        proba = new Proba(dataR.GetString(0), dataR.GetString(1), (Categorie)Enum.Parse(typeof(Categorie), dataR.GetString(3)));
                        proba.Id_arbitru = dataR.GetInt64(2);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
        }
        #endregion

        return new Rezultat(id, p, proba, scor);
    }

    public IEnumerable<Rezultat> FindAll()
    {
        log.Info("Finding all rezultate");
        var con = DBUtils.getConnection(props);
        IList<Rezultat> rezultate = new List<Rezultat>();
        try
        {
            con.Open();
            using (IDbCommand conn = con.CreateCommand())
            {
                conn.CommandText = "SELECT * FROM Rezultat";
                using (IDataReader reader = conn.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        Rezultat rezultat = fromSettoEntity(reader);
                        rezultate.Add(rezultat);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
        log.InfoFormat("Exiting FindAll with value {0}", rezultate);
        return rezultate;
    }

    public Rezultat FindOne(long id)
    {
       log.InfoFormat("Finding rezultat with id {0}", id);
        var con = DBUtils.getConnection(props);
        Rezultat rezultat = null;
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT * FROM Rezultat WHERE id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                using (IDataReader dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                         rezultat = fromSettoEntity(dataR);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
        log.InfoFormat("Exiting findOne with value {0}", rezultat);
        return rezultat;
    }

    public Dictionary<Participant, long> ParticipantiAlfabetic()
    {
        log.Info("Participanti alfabetic");
        Dictionary<Participant, long> dict = new Dictionary<Participant, long>();
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT sum(r.numar_puncte) as total_punct, p.id, p.nume,p.prenume, p.varsta " +
                    "FROM Rezultat as r Inner join Participant as p " +
                    "On r.id_participant = p.id group by r.id_participant " +
                    "order by lower(p.nume),lower(p.prenume)";
                using (IDataReader dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        log.Info("Reading data");
                        var puncte = dataR.GetInt32(0);
                        Participant par = new Participant(dataR.GetInt64(1), dataR.GetString(2), dataR.GetString(3), dataR.GetInt32(4));
                        dict.Add(par, puncte);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.Info("Exiting ParticipantScorDescrescator");
        return dict;
    }

    public Dictionary<Participant, long> ParticipantScorDescrescator(string id)
    {
        log.Info($"Proba {id} scor descrescator");
        Dictionary<Participant, long> dict = new Dictionary<Participant, long>();
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT SUM(r.numar_puncte) AS total_puncte, p.id, p.nume, p.prenume, p.varsta FROM Rezultat AS r INNER JOIN Participant AS p ON r.id_participant = p.id WHERE r.id_proba = @id GROUP BY r.id_participant ORDER BY total_puncte DESC;";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                using (IDataReader dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        log.Info("Reading data");
                        var puncte = dataR.GetInt32(0);
                        Participant par = new Participant(dataR.GetInt64(1), dataR.GetString(2), dataR.GetString(3), dataR.GetInt32(4));
                        dict.Add(par, puncte);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        dict = dict.OrderByDescending(p => p.Value).ToDictionary(p => p.Key, p => p.Value);

        log.Info("Exiting ParticipantScorDescrescator");
        return dict;
    }

    public void Save(Rezultat entity)
    {
        log.Info("Saving rezultat");
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "INSERT INTO Rezultat(id_proba,id_participant,numar_puncte) VALUES (@id_proba,@id_participant, @scor)";
                #region vars
                IDbDataParameter paramIdParticipant = comm.CreateParameter();
                paramIdParticipant.ParameterName = "@id_participant";
                paramIdParticipant.Value = entity.Participant.Id;
                comm.Parameters.Add(paramIdParticipant);

                IDbDataParameter paramIdProba = comm.CreateParameter();
                paramIdProba.ParameterName = "@id_proba";
                paramIdProba.Value = entity.Proba.Id;
                comm.Parameters.Add(paramIdProba);

                IDbDataParameter paramScor = comm.CreateParameter();
                paramScor.ParameterName = "@scor";
                paramScor.Value = entity.numar_puncte;
                comm.Parameters.Add(paramScor);
                #endregion

                var resultat = comm.ExecuteNonQuery();
                if(resultat == 0)
                {
                    log.Info("No rezultat added");
                }
                else
                {
                    log.InfoFormat("Added rezultat with value {0}", entity);
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
        log.Info("Exiting Save");
    }
    #region To implement
    public Rezultat Delete(long id)
    {
        throw new NotImplementedException();
    }


    public Rezultat Update(long id, Rezultat new_entity)
    {
        throw new NotImplementedException();
    }
    #endregion
}